node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,207 @@
+    const VALUE_TAG: u8 = 0;
+    const CHUNK_TAG: u8 = 1;
+
+    fn tag(&self) -> u8 {
+        match self {
+            ValueOrChunk::Value { .. } => Self::VALUE_TAG,
+            ValueOrChunk::ChunkWithProof(_) => Self::CHUNK_TAG,
+        }
+    }
+}
+
+impl<V: ToBytes> ToBytes for ValueOrChunk<V> {
+    fn write_bytes(&self, buf: &mut Vec<u8>) -> Result<(), bytesrepr::Error> {
+        buf.push(self.tag());
+
+        match self {
+            ValueOrChunk::Value { data, .. } => {
+                data.write_bytes(buf)?;
+            }
+            ValueOrChunk::ChunkWithProof(chunk) => {
+                chunk.write_bytes(buf)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut ret = bytesrepr::allocate_buffer(self)?;
+        self.write_bytes(&mut ret)?;
+        Ok(ret)
+    }
+
+    fn serialized_length(&self) -> usize {
+        U8_SERIALIZED_LENGTH
+            + match self {
+                ValueOrChunk::Value { data, .. } => data.serialized_length(),
+                ValueOrChunk::ChunkWithProof(chunk) => chunk.serialized_length(),
+            }
+    }
+}
+
+impl<V> FromBytes for ValueOrChunk<V> {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (tag, rem) = u8::from_bytes(bytes)?;
+        match tag {
+            Self::VALUE_TAG => {
+                let (value, rem) = Bytes::from_bytes(rem)?;
+                Ok((
+                    ValueOrChunk::Value {
+                        data: value,
+                        _phantom: PhantomData,
+                    },
+                    rem,
+                ))
+            }
+            Self::CHUNK_TAG => {
+                let (chunk, rem) = ChunkWithProof::from_bytes(rem)?;
+                Ok((ValueOrChunk::ChunkWithProof(chunk), rem))
+            }
+            _ => Err(bytesrepr::Error::Formatting),
+        }
+    }
+}
+
+/// Represents the ID of a `TrieOrChunk` - containing the index and the root hash.
+/// The root hash is the hash of the trie node as a whole.
+/// The index is the index of a chunk if the node's size is too large and requires chunking. For
+/// small nodes, it's always 0.
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct TrieOrChunkId(pub u64, pub Digest);
+
+impl TrieOrChunkId {
+    /// Returns the trie key part of the ID.
+    pub fn digest(&self) -> &Digest {
+        &self.1
+    }
+
+    /// Given a serialized ID, deserializes it for display purposes.
+    pub fn fmt_serialized(f: &mut Formatter, serialized_id: &[u8]) -> fmt::Result {
+        match bincode::deserialize::<Self>(serialized_id) {
+            Ok(ref trie_or_chunk_id) => Display::fmt(trie_or_chunk_id, f),
+            Err(_) => f.write_str("<invalid>"),
+        }
+    }
+}
+
+/// Helper struct to on-demand deserialize a trie or chunk ID for display purposes.
+pub struct TrieOrChunkIdDisplay<'a>(pub &'a [u8]);
```

Comment:
```suggestion
pub(crate) struct TrieOrChunkIdDisplay<'a>(pub &'a [u8]);
```

Reply:
Done in [`3da4c1b` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/3da4c1be1ed4e3584136bf18848f53685a90e1df)